### DISCUSSION TITLE
Community Codesandbox Gallery

### DISCUSSION THREAD

**Discussion by ebonow** [JedWatson/react-select#4341] [Show and tell]
## A place for common or tricky "how do I do this?" requests

I wanted to start a thread to perhaps give some more attention to some of the more useful implementations and feature requests out there. I understand that it would be nice if _(feature I NEED to have)_ was built-in, but there is a fine line between convenience and code bloat so thought this would be a nice place to gather some of the more useful implementations I and others have built.  This will be a running list so feel free to come back

### Editable and Selectable Text Input
When you want to always display the Input and without clearing out the inputValue on blur.
[codesandbox](https://codesandbox.io/s/react-select-editable-and-selectable-6nfdv)

### Inline Select (no borders, minimum width): 
When you want a discrete select that displays inline with other text.
[codesandbox](https://codesandbox.io/s/react-select-inline-83kdi?file=/example.js)
<img width="196" alt="Screen Shot 2020-12-20 at 5 40 24 PM" src="https://user-images.githubusercontent.com/1007579/102728761-3b0b1400-42f3-11eb-8eab-711db293fbe6.png">

### Expand width of Select to longest Option
When you want to expand the Select to the longest Option (apparently I created two different codesandboxes)
[codesandbox1](https://codesandbox.io/s/react-select-longest-option-width-geunu)
[codesandbox2](https://codesandbox.io/s/react-select-fitcontent-width-27ed6?file=/example.js)
<img width="403" alt="Screen Shot 2020-12-20 at 6 52 41 PM" src="https://user-images.githubusercontent.com/1007579/102729046-912c8700-42f4-11eb-8202-46656e51c282.png">

### Add **autoComplete** (or any other attribute) to Input
This demonstrates how to add `autoComplete` or any other Input prop attribute (ie: maxLength, required, etc...)
[codesandbox](https://codesandbox.io/s/react-select-auto-complete-2271j)

### Setting and using refs for children of custom components
This demonstrates how to approach composability to avoid creating custom components within the same scope as the Select render method.
[codesandbox](https://codesandbox.io/s/react-select-inner-inner-ref-value-container-4yuzv)

### Using IntersectionObserver to horizontally position a Menu against an edge.
This demonstrates that when a Menu is positioned against an edge and the width is set larger than the Select, it can identify whether to position and align itself to the left or right of the Select to avoid getting clipped off the screen.
[codesandbox](https://codesandbox.io/s/react-select-v3-intersection-observer-menu-position-pfly8)
<img width="900" alt="Screen Shot 2020-12-20 at 7 10 14 PM" src="https://user-images.githubusercontent.com/1007579/102729620-04cf9380-42f7-11eb-9ca4-048137f8361f.png">

### Grouping Options (Select all, identifying and styling group options)
This demonstrates how to select all options that belong to a group when clicking on a group header as well as how to style grouped options vs ungrouped options.
[codesandbox](https://codesandbox.io/s/react-select-group-option-styling-9mo4n?file=/src/App.js)
<img width="446" alt="Screen Shot 2020-12-20 at 7 36 44 PM" src="https://user-images.githubusercontent.com/1007579/102730566-bf14ca00-42fa-11eb-80d1-59bf6ada6e7b.png">

### Right-click on Select, returns context menu of input (cut, copy, paste)
This demonstrates how to make the input clickable enable users to cut, copy, paste, and select text from the text input (overriding the default 2px input width)
[codesandbox](https://codesandbox.io/s/react-select-paste-on-mobile-q51wl?file=/src/App.js)
<img width="432" alt="Screen Shot 2020-12-20 at 7 42 08 PM" src="https://user-images.githubusercontent.com/1007579/102730845-a953d480-42fb-11eb-9b04-7b86ea6a059b.png">

### Adding buttons within Menu Options
This demonstrates how to add extra functionality to your Options
[codesandbox](https://codesandbox.io/s/react-select-custom-option-gltc7?file=/example.js)
<img width="404" alt="Screen Shot 2020-12-20 at 7 56 10 PM" src="https://user-images.githubusercontent.com/1007579/102731436-77dc0880-42fd-11eb-87f2-1d4fcc1f811a.png">

### Adding styles to your scrollbar menu
This demonstrates how to add styling to the menu scrollbar
[codesandbox](https://codesandbox.io/s/react-select-custom-menu-scrollbars-c3urr)
<img width="437" alt="Screen Shot 2020-12-20 at 7 58 06 PM" src="https://user-images.githubusercontent.com/1007579/102731547-b245a580-42fd-11eb-9264-3d6afef6b66a.png">

### Displaying Menu Options in CSS Grid
This demonstrates how to display your menu as a grid
[codesandbox](https://codesandbox.io/s/react-select-grid-menu-options-ye2b8)
<img width="439" alt="Screen Shot 2020-12-20 at 7 59 19 PM" src="https://user-images.githubusercontent.com/1007579/102731612-e4ef9e00-42fd-11eb-8dcb-164e73018b49.png">

### Using CreateFilter with creatable options
This demonstrates how to properly not filter out your "Create Option" when using a custom filter with CreatableSelect.
[codesandbox](https://codesandbox.io/s/react-select-custom-filter-options-vgqcx)

### Creatable Sortable MultiValues
This demonstrates how to use the [react-sortable HOC](https://www.npmjs.com/package/react-sortable-hoc) to create sortable multiValues.
<img width="435" alt="Screen Shot 2020-12-20 at 8 06 22 PM" src="https://user-images.githubusercontent.com/1007579/102732051-3e0c0180-42ff-11eb-892b-45e6cd8ac71d.png">

### Creatable Create MultiValue on comma input
This demonstrates how to create and select new values when a comma is entered into a CreatableSelect
[codesandbox](https://codesandbox.io/s/react-select-create-on-comma-szmuw)

### Bootstrap Popover with Select
This demonstrates how to implement a Select inside of a Bootstrap Popover without prematurely closing the popover (Note use of `e.composedPath()`)
[codesandbox](https://codesandbox.io/s/react-select-v3-popover-clear-issue-composed-path-hwsbm?file=/example.js)

### Async Lookup with default option value
This demonstrates how to retrieve default options and set the default once the data has fetched
[codesandbox](https://codesandbox.io/s/react-select-async-default-option-value-clsfe?file=/example.js)
<img width="440" alt="Screen Shot 2020-12-20 at 8 21 42 PM" src="https://user-images.githubusercontent.com/1007579/102732709-0d2ccc00-4301-11eb-8c53-e62029a85e21.png">

### Shared Selects (Include/Exclude from list)
This demonstrates a POC of how to create a shared list of options (perhaps envisioning how a user could have certain roles or attributes added, or a shopping cart)
[codesandbox](https://codesandbox.io/s/react-select-include-exclude-selects-fo5uz)
<img width="438" alt="Screen Shot 2020-12-20 at 8 28 37 PM" src="https://user-images.githubusercontent.com/1007579/102733103-fe92e480-4301-11eb-9654-26a51977cabc.png">

### Input masks
This demonstrates how to write a hook to add methods to a custom input allowing which will provide the user an input mask.
[codesandbox](https://codesandbox.io/s/react-select-input-mask-4fqnf)

### MultiValues with chip overflow
This demonstrates how to show a message of setting a limit of displayed selected values with an indicator badge of how many other values are selected.
[codesandbox](https://codesandbox.io/s/custom-visible-multivalues-8zh8c?file=/example.js)
![Screen Shot 2021-10-10 at 12 00 32 PM](https://user-images.githubusercontent.com/1007579/136709673-d41546c3-5333-4b99-a5e3-4872ced500e4.png)

### MultiValues with values displayed above the Control
This demonstrates how to show the values outside of the Control and ValueContainer components.
[codesandbox](https://codesandbox.io/s/react-select-external-multi-values-98y66)
![Screen Shot 2021-10-10 at 12 03 43 PM](https://user-images.githubusercontent.com/1007579/136709745-c805016d-bb9e-4ded-b7da-0430445c889e.png)

### React-select as a HTML Select showing all options but selecting as user types
This demonstrates an example on how to recreate the behavior of a HTML Select where the values are scrolled to and selected as the user types without filtering items out of the list.
[codesandbox](https://codesandbox.io/s/react-select-native-focus-on-typed-option-show-all-options-wn9d2)
![Screen Shot 2021-10-10 at 12 07 00 PM](https://user-images.githubusercontent.com/1007579/136709837-e264a0bd-2ccf-4cdf-81d3-86d9307f911d.png)

__more examples coming__

#### Floating Labels 
_Coming soon_

#### Pillbox Multi-Select
_Coming soon_

#### Cascading Select
_Coming soon_

#### Displaying Errors
_Coming soon_

**Comment by bogdansoare** [comment]
Great collection, thanks for sharing this
